contracts/gateway/CoGateway.sol

```diff
@@ -0,0 +1,715 @@
+pragma solidity ^0.4.23;
+
+import "./MessageBus.sol";
+import "./CoreInterface.sol";
+import "./EIP20Interface.sol";
+import "./UtilityTokenAbstract.sol";
+
+contract CoGateway {
+
+	using SafeMath for uint256;
+
+	event  StakingIntentConfirmed(
+		bytes32 messageHash,
+		address staker,
+		uint256 stakerNonce,
+		address beneficiary,
+		uint256 amount,
+		uint256 blockHeight,
+		bytes32 hashLock
+	);
+
+	event MintProcessed(
+		bytes32 messageHash,
+		uint256 amount,
+		address beneficiary,
+		uint256 reward
+	);
+
+	event RevertStakingIntentConfirmed(
+		bytes32 messageHash,
+		address staker,
+		uint256 stakerNonce,
+		uint256 blockHeight
+	);
+
+	event RedeemRequested(
+		bytes32 messageHash,
+		uint256 amount,
+		address beneficiary,
+		address redeemer,
+		bytes32 intentHash
+	);
+
+	event RedeemProcessed(
+		bytes32 messageHash,
+		uint256 amount,
+		address beneficiary
+	);
+
+	event RevertRedeemRequested(
+		bytes32 messageHash,
+		address redeemer,
+		bytes32 intentHash,
+		uint256 nonce,
+		uint256 gasPrice
+	);
+
+	event RedeemReverted(
+		address redeemer,
+		uint256 amount,
+		address beneficiary,
+		uint256 gasPrice
+	);
+
+	struct Mint {
+		uint256 amount;
+		address beneficiary;
+		MessageBus.Message message;
+	}
+
+	event GatewayLinkConfirmed(
+		bytes32 messageHash,
+		address gateway,
+		address cogateway,
+		address token
+	);
+
+	/* Struct */
+	/**
+	 *  It denotes the redeem request.
+	 *  Status values could be :-
+	 *  0 :- amount used for redemption
+	 *  1 :- beneficiary is the address in the target chain where token will be minted.
```

Comment:
I don't see a status that can have a value. Generally, instead of `0` and `1` magic numbers it should use an enum (or possibly constants).